linera-service/src/chain_listener.rs

```diff
@@ -46,55 +68,114 @@ where
-            if tracker.insert(notification.clone()) {
-                info!("Received new notification: {:?}", notification);
-                if self.config.delay_before_ms > 0 {
-                    tokio::time::sleep(Duration::from_millis(self.config.delay_before_ms)).await;
-                }
-                {
-                    let mut client = self.client.lock().await;
-                    match &notification.reason {
-                        Reason::NewBlock { .. } => {
-                            if let Err(e) = client.update_validators().await {
-                                warn!(
-                                    "Failed to update validators about the local chain after \
-                                    receiving notification {:?} with error: {:?}",
-                                    notification, e
-                                );
-                            }
-                        }
-                        Reason::NewIncomingMessage { .. } => {
-                            if let Err(e) = client.process_inbox().await {
-                                warn!(
-                                    "Failed to process inbox after receiving new message: {:?} \
-                                    with error: {:?}",
-                                    notification, e
-                                );
-                            }
-                        }
-                    }
-                    context.update_wallet(client.deref_mut()).await;
-                }
+        let mut streams = HashMap::new();
+        self.update_streams(&mut streams, &mut context, &storage)
+            .await?;
+
+        while let (Some((_notification, _client)), _, _) =
+            future::select_all(streams.values_mut().map(StreamExt::next)).await
+        {}
+
+        Ok(())
+    }
 
-                if self.config.delay_after_ms > 0 {
-                    tokio::time::sleep(Duration::from_millis(self.config.delay_after_ms)).await;
+    async fn handle_notification(client: &mut ChainClient<P, S>, notification: Notification) {
+        match &notification.reason {
+            Reason::NewBlock { .. } => {
+                if let Err(e) = client.update_validators().await {
+                    warn!(
+                        "Failed to update validators about the local chain after \
+                        receiving notification {:?} with error: {:?}",
+                        notification, e
+                    );
+                }
+            }
+            Reason::NewIncomingMessage { .. } => {
+                if let Err(e) = client.process_inbox().await {
+                    warn!(
+                        "Failed to process inbox after receiving new message: {:?} \
+                        with error: {:?}",
+                        notification, e
+                    );
                 }
             }
         }
+    }
 
+    async fn update_streams<C>(
```

Comment:
I'm not sure how to do this. `update_streams` is only used in the chain listener, but the client map (I'll rename it to `ChainClients`?) is also used in the node service directly, so I don't think it would make sense there.